R/detect_column.r

```diff
@@ -0,0 +1,354 @@
+    data_sub[[i]] <- as.character(data[[i]])
+  }
+  }
+  
+  if(length(names_char_num)>0) {
+  for (j in names_char_num){
+    data_sub[[j]] <- as.numeric(data[[j]])
+  }
+  }
+  # Modify input to data at global level
+  # data_name <- deparse(substitute(data))
+  # assign(data_name,data_sub,envir=parent.frame())
+  
+  na_names <- list()
+  for(i in names(data)){
+    if(any(is.na(data[[i]]))){
+      na_names[[i]] <- i
+    }
+  }
+  na_names <- unlist(na_names)
+  if(length(na_names) > 0) {
+  stop(
+    "The following columns in dataset have NA: ",
+    paste(na_names, collapse=", ")
+  )
+  }
+  data <- data_sub
+  data
+}
+
+# validate_column_class(dat,domain_spec=domain_specs$DM)
+
+#' Check to see if columns have controlled terminology, give an error if columns do not have controlled terminology.
+#' 
+#' @param data The dataset whose columns to check.
+#' @param domain_spec The domain specification ("CO", "DM", "SE", "SM", "SV") for checking columns.
+#' @return If columns contain do not have controlled terminology, give an error and indicate which are those columns.
+#' @export
+validate_ctrl_terminology <- function(data,domain_spec){
+  if(domain_spec=="CO"){
+    data("CO")
+    domain_spec <- force(CO)
+  } else if (domain_spec=="DM"){
+    data("DM")
+    domain_spec <- force(DM)
+  } else if (domain_spec=="SE"){
+    data("SE")
+    domain_spec <- force(SE)
+  } else if (domain_spec=="SM"){
+    data("SM")
+    domain_spec <- force(SM)
+  } else if (domain_spec=="SV"){
+    data("SV")
+    domain_spec <- force(SV)
+  }
+  data("terminology")
+  terminology <- force(terminology)
+  term <- unlist(regmatches(domain_spec$'Controlled Terms, Codelist or Format', 
+                            gregexpr("(?=\\().*?(?<=\\))", 
+                                     domain_spec$'Controlled Terms, Codelist or Format', perl=T)))
+  term2 <- gsub("[()]","",term)
+  dat_term <- terminology[terminology$'CDISC Submission Value' %in% term2,]
+  dat_term2 <- terminology[terminology$'Codelist Code' %in% dat_term$Code,]
+  dat_term$`Codelist Code` <- NULL
+  names(dat_term)[names(dat_term) =='Code'] <- 'Codelist Code'
+  dat_term2 <- left_join(dat_term2,select(dat_term,'Codelist Code','CDISC Submission Value'),
+                         by='Codelist Code')
+  ctrl_variable <- domain_spec[domain_spec$'Controlled Terms, Codelist or Format' %in% term,]
+  ctrl_variable$'Controlled Terms, Codelist or Format' <- gsub("[()]","",
+                                                                 ctrl_variable$'Controlled Terms, Codelist or Format')
+  names(dat_term2)[names(dat_term2)=='CDISC Submission Value.y'] <- 'Controlled Terms, Codelist or Format'
+  dat_term2 <- left_join(dat_term2,select(ctrl_variable,'Controlled Terms, Codelist or Format',
+                                          'Variable Name'),by='Controlled Terms, Codelist or Format')
+  ctrl_variable <- ctrl_variable[ctrl_variable$'Variable Name' %in% names(data),]
```

Comment:
If I'm reading this correctly, it is getting the terminology that is allows for a particular term2 value.  I would recommend making this a different function named something like:

```
get_cdisc_controlled_terminology <- function(codelist) {
  # Verify that the codelist exists
  # Extract the values that are allowed and return them
}
```